pkg/volume/gce_pd/gce_util.go

```
@@ -122,9 +137,111 @@ func (util *GCEDiskUtil) DetachDisk(pd *gcePersistentDisk) error {
+	}
+
+	if pd.partition != "" {
+		for i, path := range devicePaths {
+			devicePaths[i] = path + "-part" + pd.partition
+		}
+	}
+	numTries := 0
+	// Wait for the disk device to detach
+	for {
+		pathExists := false
+		for _, path := range devicePaths {
+			udevadmChangeToDrive(path)
+			_, err := os.Stat(path)
+			if err == nil {
+				pathExists = true
+			}
+			if err != nil && !os.IsNotExist(err) {
+				return err
+			}
+		}
+		if !pathExists {
+			break
+		}
+		numTries++
+		if numTries == 10 {
+			return errors.New("Could not detach disk: Timeout after 10s")
+		}
+		time.Sleep(time.Second)
+	}
+
 	return nil
 }
 
+// Calls "udevadm trigger --action=change" for newly created "/dev/sd*" drives (exist only in after set).
+// This is workaround for Issue #7972. Once the underlying issue has been resolved, this may be removed.
+func udevadmChangeToNewDrives(sdBefore, sdAfter []string) {
+	sdexists := make(map[string]bool)
+	for _, sd := range sdBefore {
+		sdexists[sd] = true
+	}
+	for _, sd := range sdAfter {
+		if !sdexists[sd] {
+			udevadmChangeToDrive(sd)
+		}
+	}
+}
+
+// Calls "udevadm trigger --action=change" on the specified drive.
+// drivePath must be the the block device path to trigger on, in the format "/dev/sd*", or a symlink to it.
+// This is workaround for Issue #7972. Once the underlying issue has been resolved, this may be removed.
+func udevadmChangeToDrive(drivePath string) {
+	glog.V(5).Infof("udevadmChangeToDrive: drive=%q", drivePath)
+
+	// Evaluate symlink, if any
+	drive, err := filepath.EvalSymlinks(drivePath)
+	if err != nil {
+		glog.Errorf("udevadmChangeToDrive: filepath.EvalSymlinks(%q) failed with %v.", drivePath, err)
+		return
+	}
+	glog.V(5).Infof("udevadmChangeToDrive: symlink path is %q", drive)
+
+	// Check to make sure input is "/dev/sd*"
+	if !strings.Contains(drive, "/dev/sd") {
+		glog.Errorf("udevadmChangeToDrive: expected input in the form \"/dev/sd*\" but drive is %q.", drive)
+		return
```

You need to return an error here, not just log it.  Same for each similar case below.